encoding/time_encode.go

```diff
@@ -1,24 +1,146 @@
 package encoding
 
 import (
-	"errors"
+	"fmt"
 	"io"
+	"math"
+	"math/bits"
+	"time"
+
+	bitstream "github.com/dgryski/go-bitstream"
+	"github.com/xichen2020/eventdb/generated/proto/encodingpb"
+	"github.com/xichen2020/eventdb/x/proto"
 )
 
 // TimeEncoder encodes times values.
 type TimeEncoder interface {
 	// Encode encodes a collection of time values and writes the encoded bytes to the writer.
 	// Callers should explicitly call `Reset` before subsequent call to `Encode`.
-	Encode(writer io.Writer, values ForwardTimeIterator) error
+	Encode(writer io.Writer, values RewindableTimeIterator) error
 }
 
 // TimeEnc is a int encoder.
-type TimeEnc struct{}
+type TimeEnc struct {
+	metaProto encodingpb.TimeMeta
+	bitWriter *bitstream.BitWriter
+	buf       []byte
+}
 
 // NewTimeEncoder creates a new time encoder.
-func NewTimeEncoder(writer io.Writer) *TimeEnc { return &TimeEnc{} }
+func NewTimeEncoder() *TimeEnc {
+	return &TimeEnc{
+		bitWriter: bitstream.NewWriter(nil),
+	}
+}
 
 // Encode encodes a collection of time values and writes the encoded bytes to the writer.
-func (enc *TimeEnc) Encode(writer io.Writer, values ForwardTimeIterator) error {
-	return errors.New("not implemented")
+func (enc *TimeEnc) Encode(
+	writer io.Writer,
+	valuesIt RewindableTimeIterator,
+	resolution time.Duration,
+) error {
+	// Reset the internal state at the beginning of every `Encode` call.
+	enc.reset(writer)
+
+	var resType encodingpb.ResolutionType
+	switch resolution {
+	case time.Nanosecond:
+		resType = encodingpb.ResolutionType_NANOSECOND
+	case time.Microsecond:
+		resType = encodingpb.ResolutionType_MICROSECOND
+	case time.Millisecond:
+		resType = encodingpb.ResolutionType_MILLISECOND
+	case time.Second:
+		resType = encodingpb.ResolutionType_SECOND
+	default:
+		return fmt.Errorf("resolution (%v) is not a valid resolution", resolution)
+	}
+
+	var (
+		firstVal int64
+		max      = int64(math.MinInt64)
+		min      = int64(math.MaxInt64)
+	)
+	for idx := 0; valuesIt.Next(); idx++ {
+		curr := valuesIt.Current() / int64(resolution)
+
+		if idx == 0 {
+			firstVal = curr
+		}
+		if curr < min {
+			min = curr
+		}
+		if curr > max {
+			max = curr
+		}
+	}
+	if err := valuesIt.Err(); err != nil {
+		return err
+	}
+	valuesIt.Rewind()
+
+	enc.metaProto.Encoding = encodingpb.EncodingType_DELTA
+	enc.metaProto.Resolution = resType
+	enc.metaProto.DeltaStart = firstVal
+	enc.metaProto.BitsPerEncodedValue = int64(bits.Len(uint((max - min) + 1))) // Add 1 for the sign bit.
```

Comment:
Hm looks like it should rather be `bits.Len(uint(max - min)) + 1` instead?

Reply:
Ah, actually I don't need another bit for the sign since we don't have negative timestamps haha. 😅

Reply:
Hm but you can still have negative deltas no?

Reply:
Aren't timestamps incrementing only? Unless we are assuming out of order timestamps.

Reply:
They can be out of order since these are event timestamps that may be ingested in no particular order.

Reply:
Kk, I'll update this.